File: XMLCheck/Handler/CheckRulesXML01/Check_MA_DKBD.cs
using XMLCheck.Handler.CheckRulesXML1;
using XMLCheck.Models;
using XMLCheck.Services;

namespace XMLCheck.Handler.CheckRulesXML01
{
    public class Check_MA_DKBD : Xml1RuleBase
    {
        public override string Key => "MA_DKBD";

        public override ErrorDetails? Check(Xml1Model model, HoSoContext context)
        {
            // 1. Không được để trống
            if (string.IsNullOrWhiteSpace(model.MA_DKBD))
            {
                return Error("MA_DKBD không được để trống");
            }

            // Tách MA_DKBD
            var dkbdList = model.MA_DKBD
                .Split(';', StringSplitOptions.RemoveEmptyEntries)
                .Select(x => x.Trim())
                .ToList();

            // Rule: phải dùng dấu ;
            if (dkbdList.Count > 1 && !model.MA_DKBD.Contains(";"))
            {
                return Error("Các MA_DKBD phải cách nhau bằng dấu ';'");
            }

            // Nếu MA_THE_BHYT rỗng → không so thứ tự
            if (string.IsNullOrWhiteSpace(model.MA_THE_BHYT))
            {
                return null;
            }

            // Tách MA_THE_BHYT
            var bhytList = model.MA_THE_BHYT
                .Split(';', StringSplitOptions.RemoveEmptyEntries)
                .Select(x => x.Trim())
                .ToList();

            // 2. Số lượng phải khớp
            if (dkbdList.Count != bhytList.Count)
            {
                return Error("Thứ tự MA_DKBD không khớp với thứ tự của MA_THE_BHYT");
            }

            return null;
        }
    }
}

File: XMLValidator_v2/src/parsers/rules/xml1/Check_MA_DKBD.ts
import { Xml1RuleBase } from './xml1RuleBase';
import { Xml1Model } from '../../../interfaces/xml-models.interface';
import { HoSoContext } from '../../../services/hoso-context';
import { ErrorDetails } from '../../../dto/xml-check.dto';

export class Check_MA_DKBD extends Xml1RuleBase {
  public get key(): string {
    return 'MA_DKBD';
  }

  public check(model: Xml1Model, context: HoSoContext): ErrorDetails | null {
    if (!model.MA_DKBD || !model.MA_DKBD.trim()) {
      return this.error('MA_DKBD không được để trống');
    }

    const dkbdList = model.MA_DKBD
      .split(';')
      .map(x => x.trim())
      .filter(x => x.length > 0);

    if (dkbdList.length > 1 && !model.MA_DKBD.includes(';')) {
      return this.error("Các MA_DKBD phải cách nhau bằng dấu ';'");
    }

    if (!model.MA_THE_BHYT || !model.MA_THE_BHYT.trim()) {
      return null;
    }

    const bhytList = model.MA_THE_BHYT
      .split(';')
      .map(x => x.trim())
      .filter(x => x.length > 0);

    if (dkbdList.length !== bhytList.length) {
      return this.error("Thứ tự MA_DKBD không khớp với thứ tự của MA_THE_BHYT");
    }

    return null;
  }
}
